src/Sema.zig

```diff
@@ -12189,7 +12195,35 @@ fn zirReify(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.I
             });
             return sema.addType(ty);
         },
-        .ErrorSet => return sema.fail(block, src, "TODO: Sema.zirReify for ErrorSet", .{}),
+        .ErrorSet => switch (union_val.val.tag()) {
+            .null_value => return sema.addType(Type.initTag(.anyerror)),
+            .opt_payload => {
+                const payload_val = union_val.val.castTag(.opt_payload).?.data;
+                const slice_val = payload_val.castTag(.slice).?.data;
+                const decl = slice_val.ptr.castTag(.decl_ref).?.data;
+                try sema.ensureDeclAnalyzed(decl);
+                const array_val = decl.val.castTag(.array).?.data;
+
+                var names: Module.ErrorSet.NameMap = .{};
+                try names.ensureUnusedCapacity(sema.arena, array_val.len);
+                for (array_val) |elem_val| {
+                    const struct_val = elem_val.castTag(.@"struct").?.data;
+                    // TODO use reflection instead of magic numbers here
+                    // error_set: type,
+                    const name_val = struct_val[0];
+
+                    names.putAssumeCapacityNoClobber(
+                        try name_val.toAllocatedBytes(Type.initTag(.const_slice_u8), sema.arena),
+                        {},
+                    );
+                }
+
+                // TODO: should this be an error_set with a full decl?
+                const ty = try Type.Tag.error_set_merged.create(sema.arena, names);
+                return sema.addType(ty);
+            },
+            else => return sema.fail(block, src, "ErrorSet invalid value type", .{}),
```

Comment:
This is an internal compiler error - effectively an assertion failure - so it shouldn't use the normal compile error mechanism.

```suggestion
            else => unreachable,
```

If we wanted to make a more fail-safe build of the compiler we would probably do it by using -OReleaseSafe and having a custom panic handler that makes an error report. In such case the above `unreachable` is all it would take to trigger the appropriate mechanism.

But I think this comment becomes irrelevant after the other comment above is addressed.